extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaEightSimpleConsumerFirehoseFactory.java

```diff
@@ -0,0 +1,305 @@
+          log.error(e, "Fail to load offset from previous meta data [%s]", entry);
+        }
+      }
+      log.info("Loaded offset map[%s]", offsetMap);
+    } else {
+      log.error("Unable to cast lastCommit to Map");
+    }
+    return offsetMap;
+  }
+
+  @Override
+  public FirehoseV2 connect(final ByteBufferInputRowParser firehoseParser, Object lastCommit) throws IOException
+  {
+    final Map<Integer, Long> lastOffsets = loadOffsetFromPreviousMetaData(lastCommit);
+
+    for (Integer partition : partitionIdList) {
+      final KafkaSimpleConsumer kafkaSimpleConsumer = new KafkaSimpleConsumer(
+          feed, partition, clientId, brokerList, earliest
+      );
+      Long startOffset = lastOffsets.get(partition);
+      PartitionConsumerWorker worker = new PartitionConsumerWorker(
+          feed, kafkaSimpleConsumer, partition, startOffset == null ? 0 : startOffset
+      );
+      consumerWorkers.add(worker);
+    }
+
+    final LinkedBlockingQueue<BytesMessageWithOffset> messageQueue = new LinkedBlockingQueue<BytesMessageWithOffset>(
+        queueBufferLength
+    );
+    log.info("Kicking off all consumers");
+    for (PartitionConsumerWorker worker : consumerWorkers) {
+      worker.go(messageQueue);
+    }
+    log.info("All consumer started");
+
+    return new FirehoseV2()
+    {
+      private ConcurrentMap<Integer, Long> lastOffsetPartitions;
+      private volatile boolean stopped;
+      private volatile BytesMessageWithOffset msg = null;
+      private volatile InputRow row = null;
+
+      {
+        lastOffsetPartitions = Maps.newConcurrentMap();
+        lastOffsetPartitions.putAll(lastOffsets);
+      }
+
+      @Override
+      public void start() throws Exception
+      {
+        nextMessage();
+      }
+
+      @Override
+      public boolean advance()
+      {
+        if (stopped) {
+          return false;
+        }
+
+        nextMessage();
+        return true;
+      }
+
+      private void nextMessage()
+      {
+        try {
+          row = null;
+          while (row == null) {
+            if (msg != null) {
+              lastOffsetPartitions.put(msg.getPartition(), msg.offset());
+            }
+
+            msg = messageQueue.take();
+
+            final byte[] message = msg.message();
+            row = message == null ? null : firehoseParser.parse(ByteBuffer.wrap(message));
+          }
+        }
+        catch (InterruptedException e) {
+          //Let the caller decide whether to stop or continue when thread is interrupted.
+          log.warn(e, "Thread Interrupted while taking from queue, propagating the interrupt");
+          Thread.currentThread().interrupt();
+        }
+      }
+
+      @Override
+      public InputRow currRow()
+      {
+        if (stopped) {
+          return null;
+        }
+        return row;
+      }
+
+      @Override
+      public Committer makeCommitter()
+      {
+        final Map<Integer, Long> offsets = Maps.newHashMap(lastOffsetPartitions);
```

Comment:
This will mark the offsets from the creation of the committer rather than the Commiter.run() call. Is that intended?